Products/CMFPlone/CatalogTool.py

```diff
@@ -151,8 +151,8 @@ def allowedRolesAndUsers(obj):
 @indexer(Interface)
 def object_provides(obj):
     return tuple(
-        [i.__identifier__ for i in providedBy(obj).flattened()
-         if i.__identifier__ not in DENIED_INTERFACES]
+        i.__identifier__ for i in providedBy(obj).flattened()
+         if i.__identifier__ not in DENIED_INTERFACES
```

Comment:
Note that the previous code was faster:
```
Python 3.8.12 (default, Sep 29 2021, 09:48:53) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: my_range = tuple(range(100))

In [2]: %timeit tuple([i for i in my_range])
1.51 µs ± 3.83 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [3]: %timeit tuple(i for i in my_range)
2.42 µs ± 16.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
Strange. I simply ran `pyupgrade`.
Do you know why the old way is faster?

Reply:
This is something I learned from @jamadden in some comment of a PR in the zopefoundation organisation that I am not able to find anymore.
IIRC that was caused by some C level optimisation.

And yeah, I was surprised from that as well!

Reply:
Because it is done with less frame switching, see the first comment in:

https://stackoverflow.com/questions/43808180/python-3-x-list-comprehension-vs-tuple-generator

Reply:
```
>>> my_range = tuple(range(100))
>>> def f1():
...  tuple([i for i in my_range])
... 
>>> def f2():
...  tuple(i for i in my_range)
... 
>>> import dis
>>> dis.dis(f1)
  2           0 LOAD_GLOBAL              0 (tuple)
              3 BUILD_LIST               0
              6 LOAD_GLOBAL              1 (my_range)
              9 GET_ITER            
        >>   10 FOR_ITER                12 (to 25)
             13 STORE_FAST               0 (i)
             16 LOAD_FAST                0 (i)
             19 LIST_APPEND              2
             22 JUMP_ABSOLUTE           10
        >>   25 CALL_FUNCTION            1
             28 POP_TOP             
             29 LOAD_CONST               0 (None)
             32 RETURN_VALUE        
>>> dis.dis(f2)
  2           0 LOAD_GLOBAL              0 (tuple)
              3 LOAD_CONST               1 (<code object <genexpr> at 0x7ffbd76a31b0, file "<stdin>", line 2>)
              6 MAKE_FUNCTION            0
              9 LOAD_GLOBAL              1 (my_range)
             12 GET_ITER            
             13 CALL_FUNCTION            1
             16 CALL_FUNCTION            1
             19 POP_TOP             
             20 LOAD_CONST               0 (None)
             23 RETURN_VALUE        
>>> 
```

disassembling it shows in the second case it is using the generator, maybe the tuple conversion find it and doesn't call it until a value is requested (next())?

Reply:
> Strange. I simply ran `pyupgrade`. Do you know why the old way is faster?

I opened an issue to pyupgrade: https://github.com/asottile/pyupgrade/issues/704

Reply:
@ale-rt good catch!

Reply:
Perhaps thinking of this?

https://github.com/zopefoundation/zope.interface/blob/009c3818662a9f15696ee4b8c2beda55b1a97cb8/src/zope/interface/adapter.py#L34-L63

(https://github.com/zopefoundation/zope.interface/commit/1e720c3819b8daa07aaad0ee0258d9035597f27d)